Rewrite parallel encryption guidance (#120)
openpgp-pqc/draft-openpgp-pqc: draft-ietf-openpgp-pqc.md

 - PQ digital signature, namely SLH-DSA-SHAKE as a standalone cryptographic algorithm.

> For each of the composite schemes, this specification mandates that the recipient has to successfully perform the cryptographic algorithms for each of the component schemes used in a cryptographic message, in order for the message to be deciphered and considered as valid.

Reviewer: Since we're talking about signatures here, I guess "verifier" or something like that would work.

An alternative would be to simply state how the verification procedure works, not how an entity (recipient/comm. party) has to perform it. Something like "composite sigs are only considered valid if all component sigs are valid"

"communication party" might be weird in some contexts.

Author: I don't think it's verification only here, if we want to use `verifier` we also need to say `decrypter`?

Reviewer 2: seems to me that we have a "generating" implementation and a "consuming" or "evaluating" implementation.  "Communication party" is rather too vague.

These latter two cases realize non-composite combinations of signatures.
{{multiple-signatures}} specifies how implementations should handle the verification of such combinations of signatures.

Furthermore, the OpenPGP protocol also allows for parallel encryption to different keys.

> Accordingly, when a message is encrypted using multiple PKESK packages for different encryption keys, a non-composite multi-algorithm public-key encryption is realized where the communication party has to decrypt only one of the PKESK packages in order to decrypt the message.

Reviewer 3: This is not pertaining to the changes, but to the text as was before: Actually, whether the encryption has to be considered "multi algorithm" or not depends on whether the set of keys used belong to different algorithms. So  probably we should rather write: 
```suggestion
Accordingly, when a message is encrypted using multiple PKESK packages for different encryption keys, a non-composite, potentially multi-algorithm, public-key encryption is realized where the communication party has to decrypt only one of the PKESK packages in order to decrypt the message.
```

> Furthermore, the OpenPGP protocol also allows for parallel encryption to different keys.
> Accordingly, when a message is encrypted using multiple PKESK packets for different encryption keys, a non-composite, potentially multi-algorithm, public-key encryption is realized where the consuming party has to decrypt only one of the PKESK packets in order to decrypt the message.

Reviewer: I'd try to state this more concisely since it's more or less stating the obvious.

```suggestion
Furthermore, the OpenPGP protocol also allows parallel encryption to different keys by using multiple PKESK packets, thus realizing non-composite multi-algorithm public-key encryption.
```

> Accordingly, when a message is encrypted using multiple PKESK packages for different encryption keys, a non-composite, potentially multi-algorithm, public-key encryption is realized where the communication party has to decrypt only one of the PKESK packages in order to decrypt the message.

Reviewer 2: what is a "PKESK package"?  does this mean "packet"?  i recognize it's not a part of the text edited here, but it could be cleaned up while we're touching this line.

again, i think "communication party" is too generic.  we're talking about the "consuming" or "evaluating" implementation, not the generating implementation.  in this case, we're just talking about decryption, so we could just say "the decrypting party".

# Algorithm Combinations

## Composite KEMs {#composite-kem}

The ML-KEM + ECDH public-key encryption involves both the ML-KEM and an ECDH KEM in an a priori non-separable manner.
This is achieved via KEM combination, i.e. both key encapsulations/decapsulations are performed in parallel, and the resulting key shares are fed into a key combiner to produce a single shared secret for message encryption.

As explained in {{non-composite-multi-alg}}, the OpenPGP protocol inherently supports parallel encryption to different keys.
Note that the confidentiality of a message is not post-quantum secure when encrypting to different keys if at least one key does not support PQ/T encryption schemes.

> In section {{pq-key-preference}} is explained how to deal with multiple key scenarios.

Reviewer 3: ```suggestion
In section {{pq-key-preference}} it is explained how to deal with multiple key scenarios.
```

## Symmetric Algorithms for SEIPD Packets

Implementations MUST implement `AES-256`.

> An implementation SHOULD use `AES-256` in the case of a v1 SEIPD packet, or `AES-256` with any available AEAD mode in the case of a v2 SEIPD packet, if all communication parties indicate support for it (explicitly or implicitly).

Reviewer: I think the feedback in the Interim was not to avoid the word "recipient" everywhere. I think here it is sufficiently clear what is meant by "recipient" and I think it's the best word to use. The ambiguity comes in when matching multiple keys/certs to a recipient / entity and it becomes a more abstract concept.

For a PKESK packet there is a recipient / recipient key and for a SEIPD message the recipients are each recipient of the list of PKESK packets. I think that's common wording also in the Crypto Refresh.

Reviewer 3: Yes, it is correct the the term "recipient" is used throughout the crypto-refresh. So we should rather stick to it and probably just avoid conveying the idea of differentiating between keys that belong to a single recipient and to multiple recipients.

Author: ```suggestion
An implementation SHOULD use `AES-256` in the case of a v1 SEIPD packet, or `AES-256` with any available AEAD mode in the case of a v2 SEIPD packet, if all recipients indicate support for it (explicitly or implicitly).
```

During the transition period, the post-quantum algorithms will not be supported by all clients.
Therefore various migration considerations must be taken into account, in particular backwards compatibility to existing implementations that have not yet been updated to support the post-quantum algorithms.

## Key preference {#pq-key-preference}

Implementations SHOULD prefer PQ(/T) keys when multiple options are available.

> In case that a certificate has both a valid PQ/T and a valid traditional subkey, an implementation SHOULD use the PQ/T subkey only.

Reviewer: Sorry for the holdup, but I just noticed that we should phrase it a bit narrower. Otherwise this statement also includes the signature case and contradicts what we write in line 905, 906.

```suggestion
When encrypting to a certificate that has both a valid PQ/T and a valid traditional encryption subkey, an implementation SHOULD use the PQ/T subkey only.
```

Furthermore, if an application has any means to determine that encrypting to a PQ/T certificate and a traditional certificate is redundant, it should omit encrypting to the traditional certificate.

As specified in {{composite-kem}}, the confidentiality of a message is not post-quantum secure when using multiple PKESKs if at least one does not use PQ/T encryption schemes.
An implementation SHOULD NOT abort the encryption process when encrypting a message to both PQ/T and traditional keys to allow for a smooth transition to post-quantum cryptography.

> In case a certificate has both a PQ/T and an ECC encryption-capable valid subkey, the PQ/T subkey SHOULD be preferred.

Reviewer: still needs to be removed

> In case a certificate has both a PQ/T and an ECC encryption-capable valid subkey, the PQ/T subkey SHOULD be preferred.

Reviewer: ```suggestion
```

During the transition period, the post-quantum algorithms will not be supported by all clients.
Therefore various migration considerations must be taken into account, in particular backwards compatibility to existing implementations that have not yet been updated to support the post-quantum algorithms.

## Key preference {#pq-key-preference}

Implementations SHOULD prefer PQ(/T) keys when multiple options are available.

> For instance, an implementation that encrypts messages for a recipient for which both a valid PQ/T and a valid ECC certificate are available, the implementation should use the PQ/T certificate only.

Reviewer: Here I think we run into the problem that the abstract concept of a recipient (outside of PKESKs) is not well defined. Might be ok for the non-normative statement, though.

Author: I think it's fine as an explaination of the previous sentence

Reviewer: I would still rather rewrite this. IMO that problem is outside of the scope of the OpenPGP protocol and more an application thing. The protocol layer doesn't necessarily know that two certificates / keys that it processes for encrypting are somehow linked. And even if it does, it can't know the intent of the application/user.
See also #67, I think the critique still applies to this: Do we mean the second or the third scenario, or both, and what is the criterion for an application to determine that two certificates belong to the same recipient in the third scenario?

In any case, I would change the paragraph to start with the more obvious case first:

> For instance, in the case that a certificate has both a PQ/T and an ECC encryption-capable valid subkey, the PQ/T subkey SHOULD be preferred.

And then we can add the "multiple certificate scenario" a bit more vaguely. The best I could come up with is the following:

> Further, if an application can determine (by any means not defined in this document) that encrypting to a PQ/T certificate and a traditional certificate serves the same purpose, it should omit encrypting to the traditional certificate.

Sorry for being a bit picky here but I think sooner or later someone will re-raise this issue.

Reviewer 3: After looking closer again I agree with Johannes, here the assumptions about the identity of the recipient will most likely have to be more vague.

Author: ```suggestion
For instance, in the case that a certificate has both a PQ/T and an ECC encryption-capable valid subkey, the PQ/T subkey SHOULD be preferred.
Furthermore, if an application can anyhow determine that encrypting to a PQ/T certificate and a traditional certificate is redundant, it should omit encrypting to the traditional certificate.
```
What about this?

Reviewer 3: I think we're almost there. But "SHOULD be preferred" might not be clear enough, as we also want to express that in this case the traditional key should not be used at all. So I would write for instance " ... valid subkey, the PQ/T subkey SHOULD be used and the traditional key SHOULD NOT be used."

Furthermore, I wouldn't write "and an *ECC* encryption-capable valid subkey" but rather "and a *traditional* encryption-capable valid subkey" in order not to create the false assumption that it only refers to ECC keys (other types of keys might be around for quite some time).

Author: That's the point of "for instance", IMO. I would rather remove the normative part from an example then or remove the "for instance".

Normative way:
```suggestion
In case that a certificate has both a valid PQ/T and a valid traditional subkey, an implementation SHOULD use the PQ/T subkey only.
Furthermore, if an application can anyhow determine that encrypting to a PQ/T certificate and a traditional certificate is redundant, it should omit encrypting to the traditional certificate.
```
Example way:
```suggestion
For instance, an implementation that encrypts messages for a recipient for which both a valid PQ/T and a valid ECC encryption subkey are available, the implementation should use the PQ/T subkey only.
Furthermore, if an application can anyhow determine that encrypting to a PQ/T certificate and a traditional certificate is redundant, it should omit encrypting to the traditional certificate.
```
Somehow hybrid, but not normative:
```suggestion
For instance, an implementation that encrypts messages for a recipient for which both a valid PQ/T and a valid traditional encryption subkey are available, the implementation should use the PQ/T subkey only.
Furthermore, if an application can anyhow determine that encrypting to a PQ/T certificate and a traditional certificate is redundant, it should omit encrypting to the traditional certificate.
```

Reviewer: I would write:

> For instance, an implementation that encrypts a message to a certificate for which both a valid PQ/T and a valid traditional encryption subkey are available, the implementation should encrypt to the PQ/T subkey only.

This is adapted from your "Example way" to avoid the "recipient" terminology, and I also replaced "use" by "encrypt to", as well as "ECC" by "traditional".

I'm happy with your suggestion for the "multiple certificate scenario".

Reviewer 3: I vote for the normative variant @wussler has posted above.

(An example with ECC is in my point potentially misleaing – even though from a formal point it is correct since it is an example. But our aim here by rendering the guidance as an example is only with respect to the set of keys, I don't see a need to make it concrete with respect to the public-key encryption algorithm.)

I also generally agree to the second sentence about the redundant certificate. That solves the problem of the using the word recipient well in my view. But the use of the word "anyhow" doesn't seem right to me. I ran the sentence through deepl and it doesn't translate to German correctly. I think we need to replace "anyhow" with "in some way" or with "somehow". Or maybe even better 

"Furthermore, if an application has any means to determine ..."

Reviewer: Ok, then let's go with the normative variant.

I was a bit unsure about "anyhow" myself but cannot judge it. I vote for your suggestion to write "Furthermore, if an application has any means to determine ...".

For instance, an implementation that encrypts messages for a recipient for which both a valid PQ/T and a valid ECC certificate are available, the implementation should use the PQ/T certificate only.

As specified in {{composite-kem}}, the confidentiality of a message is not post-quantum secure when using multiple PKESKs if at least one does not use PQ/T encryption schemes.
An implementation SHOULD NOT abort the encryption process when encrypting a message to both PQ/T and traditional keys to allow for a smooth transition to post-quantum cryptography.

For instance, if encrypting for a recipient for which both a valid PQ/T and a valid ECC certificate are available, the implementation SHOULD choose the PQ/T certificate.
In case a certificate has both a PQ/T and an ECC encryption-capable valid subkey, the PQ/T subkey SHOULD be preferred.

An implementation MAY sign with both a PQ(/T) and an ECC key using multiple signatures over the same data as described in {{multiple-signatures}}.
Signing only with PQ(/T) key material is not backwards compatible.

> Note that the confidentiality of a message is not post-quantum secure when encrypting to multiple recipients if at least one recipient does not support PQ/T encryption schemes.

Reviewer: I think that here the word "recipient" would be unterstood in the context of the PKESK packet(s) so it'd be ok to use

## Security Aspects of Composite Signatures

When multiple signatures are applied to a message, the question of the protocol's resistance against signature stripping attacks naturally arises.

> In a signature stripping attack, an adversary removes one or more of the signatures such that only a subset of the signatures remain in the message when verified.

Reviewer 3: ```suggestion
In a signature stripping attack, an adversary removes one or more of the signatures such that only a subset of the signatures remain in the message at the point when it is verified.
```

First, `AES-KeyWrap` with `AES-256` is already part of the composite KEM construction.
Second, some of the PQ(/T) algorithms target the security level of `AES-256`.

> For the same reasons, this specification further recommends the use of `AES-256` if it is supported by all communication parties, regardless of what the implementation would otherwise choose based on the communication parties' preferences.

Reviewer: Also here I think "recipient" is ok to use
